accepted/0000-hub-online-synchronization.md

```diff
@@ -0,0 +1,228 @@
+
+### HUB XML-RPC API
+
+Component responsible for broadcasting API requests to peripheral servers. This component loads the peripheral servers that are registered on the HUB by looking at the minion entitlements.
+
+With the proposed solution we can retrieve the list of the registered peripheral servers and avoid the peripheral server registration as minion.
+
+### Inter-server-sync-v2
+
+Will be replaced by this solution when fully implemented (including images and configuration channels). It will continue to exist but for disconnect set-ups only.
+
+### Uyuni-config formula
+
+This proposal doesn't address any of the issues with the configuration formula.
+
+### HUB Report Federation
+
+Reporting database information is collected using a salt state to get the report DB connection information.
+With this solution we can replace the mechanism by an API call to the peripheral server which would return the needed information, and in this way remove the need to minion registration.
+
+
+## Future Steps
+
+This implementation sets the foundations to add more online content synchronization through the API, like configuration channels, images or activation keys.
+this section will define a initial approach to make it possible. It's not as details as the proposed solution, since it's out of implementation scoped for this RFC and is described to give a full solution picture.
+
+### Configuration channels
+
+All data for configuration channel are save in the database and then materialized in disk to be used by salt. The source of true is the database information and the disk materialization is automatic if API methods are used.
+
+In the HUB server we can collect which configuration channels that should be synchronized to each peripheral. Configuration channel are attached to a organization and they should be part of the organization mapping selection.
+It's label behave in a different way when compared to software channels: we can have the same label in different organizations.
+
+A new taskomatic Job can be created to synchronized data from HUB server to peripherals which should run automatically once per day during a low usage time.
+
+The existing API methods under namespace `configchannel` cannot deal with cross organization mapping. To overcome it we have two options:
+   - Develop a new endpoint to create and update channels with server-to-server authentication and the possibility to defined the target organization. Besides organization identifier, all business logic should be the same as the API namespace `configchannel`.
+   - Collect organization admin credentials (user+password) on the HUB, and use the existing API. The downside is that it's more fields to be collected and managed.
+
+### Images
+
+TODO
+- Is all data saved in the database?
+- Is the identifier unique in the system, or we can have duplicates across organizations?
+- Are the existing endpoint sufficient to update all the data?
```

Comment:
> Is all data saved in the database?

Metadata + pillar data are in the database, images are separate, http(s) accessible

> Is the identifier unique in the system, or we can have duplicates across organizations?

Image IDs are unique across the installation.

> Are the existing endpoint sufficient to update all the data?

Split over multiple API calls + http for image download is there.
Importing the image is lacking. Only basic information + pillar + image files can be imported, e.g. package list is not possible to import using API.

Example implementation is in [contrip/os-image-tools](https://github.com/uyuni-project/contrib/blob/main/os-image-tools/osimage-import-export.py)

From my POV, once the API authentication using the token is available, the exporting of images is ready. Only thing needed is that peripheral server regularly calls hub server API to ask for new or removed images.
Some performance optimizations can be done, like time based image search and adding more complex API calls to limit needed number of calls can be done.

For import, this is quite lacking:
- image building handling is mangled together with salt event handling -> quite a refactoring to do so
- once refactored, we can reuse same classes and methods to do import of images + add an API endpoints.

Reply:
Thank you Ondrej
I have added the information about images.